services/triggerService.ts

```diff
@@ -187,16 +189,46 @@ export async function executeAddressTriggers (broadcastTxData: BroadcastTxData):
       paymentAddress,
       timestamp
     }
-    await postDataForTrigger(trigger, postDataParameters)
+    const postHash = await hashPostData(trigger.paybutton.providerUserId, postDataParameters)
+    await postDataForTrigger(trigger, {
+      ...postDataParameters,
+      postHash
+    }
+    )
   }))
 }
 
-async function postDataForTrigger (trigger: TriggerWithPaybutton, postDataParameters: PostDataParameters): Promise<void> {
+export interface PostDataParameters {
+  amount: Prisma.Decimal
+  currency: string
+  timestamp: number
+  txId: string
+  buttonName: string
+  paymentAddress: string
+}
+
+export interface PostDataParametersHashed {
+  amount: Prisma.Decimal
+  currency: string
+  timestamp: number
+  txId: string
+  buttonName: string
+  paymentAddress: string
+  postHash: string
+}
+
+async function hashPostData (userId: string, postDataParameters: PostDataParameters): Promise<string> {
+  const hmac = crypto.createHmac('sha256', await getUserSecretKey(userId))
+  hmac.update(`${postDataParameters.timestamp}}`)
```

Comment:
We want to hash all of the data we're sending so let's split this out into two steps:

1. Hash all of the data (however you want to package it) using vanilla SHA256.
2. Hash / encrypt / sign the hash with SHA256 again but this time with the user secret.

This way we get the benefit of it being easy to verify as it's still just one string but also are hashing all of the data since there's a chance of duplicates if 2 txs come in at the same time.

Reply:
Good point, I addressed it hashing all ` amount, currency, paymentAddress, timestamp, txId ` but not `buttonName` because that could change and lead to problems